### PR TITLE
LPS-147229 Review ContentDashboardItem web API for the Sidebar Info Panel

### PR DESCRIPTION
**Pull request by cgoncas**
Motivation
=======
We are sending vía REST from the backend to the frontend a not well structured object. We are repeating values (ie. dates) and the object contains two similar properties data and specificFields, used for the same purpose of sending item details.

On the frontend side, we render the info manually attending for each specific property and assigning the right component.
We aim to do this programmatically, using a more complex structure of values from the backend that can lead to a smarter rendering in the frontend.

[Link to story or ticket](https://issues.liferay.com/browse/LPS-147229)

Proposed Solution
========

In the backend side:
- Create an object Copy in the ContentDashboardItem api that represents this object in the UI:

Screenshot 2022-02-21 at 12 39 36

- Create an object Preview in the ContentDashboardItem api that represents this object in the UI:
Screenshot 2022-02-21 at 12 40 50

- The final format that the backend will send to the frontend will be:
```
{
  "preview": {
    "imageURL": "http://localhost:8080/documents/20123/0/foto.jpeg/f3e5fe66-961d-f164-38b7-f3545bad2f10?t=1645177483072",
    "downloadURL": "/documents/20123/0/foto.jpeg/f3e5fe66-961d-f164-38b7-f3545bad2f10?version=1.0&t=1645177483072&download=true",
    "url": "http://localhost:8080/group/guest/~/control_panel/manage?p_p_id=com_liferay_document_library_web_portlet_DLAdminPortlet&p_p_lifecycle=0&p_p_state=maximized&_com_liferay_document_library_web_portlet_DLAdminPortlet_mvcRenderCommandName=%2Fdocument_library%2Fview_file_entry&_com_liferay_document_library_web_portlet_DLAdminPortlet_fileEntryId=42520&p_p_auth=hRvwEH6a&_com_liferay_document_library_web_portlet_DLAdminPortlet_redirect=%2Fgroup%2Fguest%2F%7E%2Fcontrol_panel%2Fmanage%3Fp_p_id%3Dcom_liferay_content_dashboard_web_portlet_ContentDashboardAdminPortlet%26p_p_lifecycle%3D0%26p_p_state%3Dmaximized%26p_v_l_s_g_id%3D20123%26p_p_auth%3DsxLqJ1Tq"
  },
  "vocabularies": {},
  "specificFields": {
    "extension": {
      "title": "Extension",
      "type": "string",
      "value": "jpeg"
    },
    "size": {
      "title": "Size",
      "type": "string",
      "value": "9 KB"
    }
  },
  "description": "",
  "className": "com.liferay.portal.kernel.repository.model.FileEntry",
  "languageTag": "en-US",
  "title": "foto",
  "type": "Document",
  "viewURLs": [
    {
      "default": true,
      "languageId": "en-US"
    }
  ],
  "tags": [],
  "classPK": "42520",
  "versions": [
    {
      "statusStyle": "success",
      "version": "1.0",
      "statusLabel": "Approved"
    }
  ],
  "modifiedDate": "2022-02-18T09:44:43.072",
  "subType": "Basic Document",
  "copy": {
    "name": "foto.jpeg",
    "url": "http://localhost:8080/documents/20123/0/foto.jpeg/f3e5fe66-961d-f164-38b7-f3545bad2f10?t=1645177483072"
  },
  "user": {
    "name": "Test Test",
    "userId": "20127"
  },
  "createDate": "2022-02-18T09:44:42.979"
}
```

This is just a proposal, please fell free to change everything, I will create separate commits and tests ASAP, but I would want you to take a first look ASAP.

Steps to Verify:
----------------
 NO FUNCTIONAL CHANGES

 All Content Dashboard tests (manual and automated) should work

### CODE REVIEW

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:tango

**Review by beltranrengifo — COMMENTED** [review]
Hi @cgoncas, thanks for this proposal! I think it's better this way however I left you some comments:

- Can we have a similar example for a web content? Just to evaluate is everything fit in
- the _Copy_ component needs just the url, the fileName is rendered separately and I think is better that way
- As the fileName is specific for documents, make sense to place this field inside the specificFields object?

**Comment by cgoncas** [comment]
> Hi @cgoncas, thanks for this proposal! I think it's better this way however I left you some comments:
> 
> * Can we have a similar example for a web content? Just to evaluate is everything fit in
> * the _Copy_ component needs just the url, the fileName is rendered separately and I think is better that way
> * As the fileName is specific for documents, make sense to place this field inside the specificFields object?

Hey @beltranrengifo !!

But the fileName is only shown as an adding information for the URL you are trying to copy, right? It is not a specific field, because it is shown as the subtitle of the copy URL.

This is the WebContet  example:


Screenshot 2022-02-21 at 10 25 42

Thanks!!

**Comment by beltranrengifo** [comment]
> 



> But the fileName is only shown as an adding information for the URL you are trying to copy, right? It is not a specific field, because it is shown as the subtitle of the copy URL.

Uhmm yes, you are right, that value is not used anywhere else.

Then, should we proceed to adjust the frontend in this PR?

**Comment by cgoncas** [comment]
> > 
> 
> > But the fileName is only shown as an adding information for the URL you are trying to copy, right? It is not a specific field, because it is shown as the subtitle of the copy URL.
> 
> Uhmm yes, you are right, that value is not used anywhere else.
> 
> Then, should we proceed to adjust the frontend in this PR?

Yeahh!!!

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java [hunk 1]
```diff
@@ -404,6 +396,32 @@ public boolean isViewable(HttpServletRequest httpServletRequest) {
...
+						ContentDashboardItemAction.Type.PREVIEW);
+
+				if (!contentDashboardItemActions.isEmpty()) {
+					ContentDashboardItemAction contentDashboardItemAction =
+						contentDashboardItemActions.get(0);
+
+					return contentDashboardItemAction.getURL();
+				}
+
+				return null;
```
Comment (YurenaC) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java:418]: Just wondering if those "null" can cause any NullPointerException in other parts of the code
Maybe return empty string is safer?

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 2]
```diff
@@ -130,20 +131,23 @@ protected void doServeResource(
 					"className", _getClassName(contentDashboardItem)
 				).put(
 					"classPK", _getClassPK(contentDashboardItem)
+				).put(
+					"copy", _getCopyJSONObject(contentDashboardItem)
```
Comment (YurenaC) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:135]: Why do we need this copy?

Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:135]: copy is the name of the FE component we use to copy the file URL
image

Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:135]: But if you think we can use a better name, lest rename it :)

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java (hunk 1 above)
Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java:418]: A null value is translated to a no value, but the empty string is used as is the actual string is empty. Where do you think a null pointer exception can be thrown?

Comment (YurenaC) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java:418]: just checked all places where is used and won't produce any exception.
Nicely done @cgoncas  ;)

**Review by claraizquierdo — CHANGES_REQUESTED** [review]
Copy url should not be sent when the item is an internal video shortcut.

**Review by claraizquierdo — CHANGES_REQUESTED** [review]
Download button should not be present in External Video Shortcut
Captura de pantalla 2022-03-01 a las 12 48 02

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
Hey @claraizquierdo !!

I have fixed the Download URL for the External Video Shortcut, but I would need the exact list of what are the documents that support Download, and what are the documents that support Preview!!

Thanks!!

**Comment by claraizquierdo** [comment]
> Copy url should not be sent when the item is an internal video shortcut.

@claraizquierdo I guess that we are going to fix this in this task right https://issues.liferay.com/browse/LPS-131186 ?

**Comment by claraizquierdo** [comment]
ci:test:relevant

**Review by claraizquierdo — APPROVED** [review]

**Comment by YurenaC** [comment]
> Hey @YurenaC !!
> 
> Could you help my to find with the XLS test is broken?
> 
> Thanks!!

Hi Cris!
The "File Name" field is missing in the generated excel file.

I debugged a little bit and in the file "GetContentDashboardItemsXlsMVCResourceCommand" line 199 (_toString(specificInformation, "fileName")), it's returning empty as "fileName" is not in the specificInformation I think

**Comment by cgoncas** [comment]
> > Hey @YurenaC !!
> > Could you help my to find with the XLS test is broken?
> > Thanks!!
> 
> Hi @cgoncas ! The "File Name" field is missing in the generated excel file.
> 
> I debugged a little bit and in the file "GetContentDashboardItemsXlsMVCResourceCommand" line 199 (_toString(specificInformation, "fileName")), it's returning empty as "fileName" is not in the specificInformation I think

Awesome @YurenaC !!!

Thanks a lot, I will fix it ASA :)

**Comment by cgoncas** [comment]
> > > Hey @YurenaC !!
> > > Could you help my to find with the XLS test is broken?
> > > Thanks!!
> > 
> > 
> > Hi @cgoncas ! The "File Name" field is missing in the generated excel file.
> > I debugged a little bit and in the file "GetContentDashboardItemsXlsMVCResourceCommand" line 199 (_toString(specificInformation, "fileName")), it's returning empty as "fileName" is not in the specificInformation I think
> 
> Awesome @YurenaC !!!
> 
> Thanks a lot, I will fix it ASAP :)

I think I just fix it, I will start the testing, to be sure :)

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:reevaluate:1347401_1028

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
Hey @YurenaC !!

I am deleting the XLS integration test, because I have added the creation date and other fields that are hard to test, so I will move it to a unit test, it is hard to do it without a refactor, so I will send this PR, and we can work in the test later.


Thanks!!!

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:forward

**Pull request closed without merging**